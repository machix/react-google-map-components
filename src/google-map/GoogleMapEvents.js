export const ON_MOUSE_MOVE = "mousemove";

export const ON_IDLE = "idle";
export const ON_TILES_LOADED = "tilesloaded";
export const ON_TILT_CHANGED = "tilt_changed";
export const ON_ZOOM_CHANGED = "zoom_changed";
export const ON_BOUNDS_CHANGED = "bounds_changed";
export const ON_CENTER_CHANGED = "center_changed";
export const ON_HEADING_CHANGED = "heading_changed";
export const ON_MAP_TYPE_ID_CHANGED = "maptypeid_changed";
export const ON_PROJECTION_CHANGED = "projection_changed";
