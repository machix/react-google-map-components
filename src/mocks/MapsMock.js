export function createMapsMock() {
  return {
    Map: jest.fn(function GoogleMapsMap(node, options) {
      this.node = node;
      this.options = options;
      this.listeners = {};

      this.setValues = jest.fn();
      this.addListener = jest.fn((event, listener) => {
        this.listeners[event] = this.listeners[event] || [];
        this.listeners[event].push(listener);
      });
    }),

    Size: jest.fn(),
    Point: jest.fn(),
    LatLng: jest.fn(),

    LatLngBounds: jest.fn(function GoogleMapsLatLngBounds() {
      this.extend = () => jest.fn();
    }),

    event: {
      clearInstanceListeners: jest.fn(),
    },
  };
}
