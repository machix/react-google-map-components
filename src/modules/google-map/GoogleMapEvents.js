import GenericEvents from "../internal/GenericEvents";

export default {
  onClick: GenericEvents.onClick,
  onDoubleClick: GenericEvents.onDoubleClick,
  onRightClick: GenericEvents.onRightClick,
  onMouseOut: GenericEvents.onMouseOut,
  onMouseOver: GenericEvents.onMouseOver,
  onMouseMove: GenericEvents.onMouseMove,
  onDrag: GenericEvents.onDrag,
  onDragStart: GenericEvents.onDragStart,
  onDragEnd: GenericEvents.onDragEnd,

  onIdle: "idle",
  onTilesLoaded: "tilesloaded",
  onTiltChanged: "tilt_changed",
  onZoomChanged: "zoom_changed",
  onBoundsChanged: "bounds_changed",
  onCenterChanged: "center_changed",
  onHeadingChanged: "heading_changed",
  onMapTypeIdChanged: "maptypeid_changed",
  onProjectionChanged: "projection_changed",
};
