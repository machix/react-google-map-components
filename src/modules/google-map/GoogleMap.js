import React from "react";
import _ from "lodash";
import PropTypes from "prop-types";
import GoogleMapEvents from "./GoogleMapEvents";
import { createListeners, getChangedProps } from "../internal/Utils";
import { MapContext } from "../internal/MapContext";

const styles = { map: { height: "100%" } };

const pickProps = x =>
  _.pick(x, [
    "zoom",
    "center",
    "mapTypeId",
    "clickableIcons",
    "backgroundColor",
    "disableDoubleClickZoom",
  ]);

/**
 * Draws `google.maps.Map`.
 *
 * **Usage:**
 *
 * ```javascript
 * import React from "react";
 * import { GoogleMap } from "react-google-map-components"
 *
 * export default function GoogleMapWrapper(props) {
 *   return <GoogleMap {...props} maps={google.maps} />
 * }
 * ```
 *
 * **Google Maps Docs:**
 * * [google.maps.Map](https://developers.google.com/maps/documentation/javascript/reference#Map)
 * * [google.maps.MapOptions](https://developers.google.com/maps/documentation/javascript/reference#MapOptions)
 */
export class GoogleMap extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.map = null;
    this.node = null;
    this.maps = props.maps;
    this.mapContext = null;
  }

  getChildContext() {
    return { mapContext: this.mapContext };
  }

  componentDidMount() {
    this.map = new this.maps.Map(this.node);
    this.mapContext = new MapContext(this.map, this.maps);

    const options = this.getOptions(this.props);

    options.disableDefaultUI = true;

    this.map.setValues(options);

    createListeners(GoogleMapEvents, x => this.props[x]).forEach(
      ([event, listener]) => {
        switch (event) {
          case GoogleMapEvents.onBoundsChanged:
            this.map.addListener(event, () =>
              listener({ bounds: this.map.getBounds() }),
            );
            break;

          default:
            this.map.addListener(event, listener);
        }
      },
    );

    this.forceUpdate();
  }

  componentDidUpdate(prevProps) {
    const prevOptions = this.getOptions(prevProps);
    const nextOptions = this.getOptions(this.props);
    const options = getChangedProps(prevOptions, nextOptions);

    if (options) {
      this.map.setValues(options);
    }
  }

  componentWillUnmount() {
    this.maps.event.clearInstanceListeners(this.map);
  }

  getOptions(props) {
    const options = pickProps(props);

    options.mapTypeId = this.mapContext.getEnum("MapTypeId", options.mapTypeId);

    return options;
  }

  render() {
    const { style, className } = this.props;

    return (
      <div style={style} className={className}>
        <div style={styles.map} ref={x => (this.node = x)} />

        {Boolean(this.mapContext) && this.props.children}
      </div>
    );
  }
}

GoogleMap.childContextTypes = {
  mapContext: PropTypes.instanceOf(MapContext),
};

GoogleMap.defaultProps = {
  mapTypeId: "ROADMAP",

  clickableIcons: true,
  disableDoubleClickZoom: false,
};

/* istanbul ignore else */
if (process.env.NODE_ENV !== "production") {
  GoogleMap.propTypes = {
    /**
     * Loaded `google.Maps` instance.
     */
    maps: PropTypes.object.isRequired,

    /**
     * The initial Map `center`.
     */
    center: PropTypes.shape({
      /**
       * The latitude in degrees.
       */
      lat: PropTypes.number.isRequired,
      /**
       * The longitude in degrees.
       */
      lng: PropTypes.number.isRequired,
    }).isRequired,

    /**
     * The initial Map `zoom` level.
     */
    zoom: PropTypes.number.isRequired,

    /**
     * The initial Map `mapTypeId`.
     */
    mapTypeId: PropTypes.oneOf(["HYBRID", "ROADMAP", "SATELLITE", "TERRAIN"]),

    /**
     * Google Maps child components.
     */
    children: PropTypes.node,

    /**
     * When `false`, map icons are not clickable.
     *
     * A map icon represents a point of interest, also known as a POI.
     */
    clickableIcons: PropTypes.bool,

    /**
     * Enables/disables `zoom` and `center` on double click.
     */
    disableDoubleClickZoom: PropTypes.bool,

    //
    // Styling
    //

    /**
     * Styles of map div.
     */
    style: PropTypes.object,

    /**
     * Classes of map div.
     */
    className: PropTypes.string,

    /**
     * Color used for the background of the Map.
     *
     * This color will be visible when tiles have not yet
     * loaded as the user pans.
     */
    backgroundColor: PropTypes.string,

    //
    // Events
    //

    /**
     * This handler is called when the user clicks on the map.
     *
     * An `ApiMouseEvent` with properties for the clicked location is returned
     * unless a place icon was clicked, in which case an `IconMouseEvent`
     * with a `placeid` is returned.
     *
     * `IconMouseEvent` and `ApiMouseEvent` are identical, except that
     * `IconMouseEvent` has the `placeid` field.
     *
     * The event can always be treated as an `ApiMouseEvent` when the
     * `placeid` is not important.
     *
     * The click event is not fired if a `Marker` or `InfoWindow` was clicked.
     */
    onClick: PropTypes.func,

    /**
     * This handler is called when the user double-clicks on the map.
     *
     * Note that the `onClick` handler will be also called,
     * right before this one.
     */
    onDoubleClick: PropTypes.func,

    /**
     * This handler is called when the DOM context menu event is
     * fired on the map container.
     */
    onRightClick: PropTypes.func,

    /**
     * This handler is called whenever the user's mouse moves over
     * the map container.
     */
    onMouseMove: PropTypes.func,

    /**
     * This handler is called when the user's mouse exits the map container.
     */
    onMouseOut: PropTypes.func,

    /**
     * This handler is called when the user's mouse enters the map container.
     */
    onMouseOver: PropTypes.func,

    /**
     * This handler is repeatedly called while the user drags the map.
     */
    onDrag: PropTypes.func,

    /**
     * This handler is called when the user stops dragging the map.
     */
    onDragEnd: PropTypes.func,

    /**
     * This handler is called when the user starts dragging the map.
     */
    onDragStart: PropTypes.func,

    /**
     * This handler is called when the map becomes idle after panning or zooming.
     */
    onIdle: PropTypes.func,

    /**
     * This handler is called when the visible tiles have finished loading.
     */
    onTilesLoaded: PropTypes.func,

    /**
     * This handler is called when the map `tilt` property changes.
     */
    onTiltChanged: PropTypes.func,

    /**
     * This event is fired when the map `zoom` property changes.
     */
    onZoomChanged: PropTypes.func,

    /**
     * This handler is called when the viewport bounds have changed.
     */
    onBoundsChanged: PropTypes.func,

    /**
     * This handler is called when the map `center` property changes.
     */
    onCenterChanged: PropTypes.func,

    /**
     * This handler is called when the map `heading` property changes.
     */
    onHeadingChanged: PropTypes.func,

    /**
     * This handler is called when the `mapTypeId` property changes.
     */
    onMapTypeIdChanged: PropTypes.func,

    /**
     * This handler is called when the `projection` has changed.
     */
    onProjectionChanged: PropTypes.func,
  };
}
