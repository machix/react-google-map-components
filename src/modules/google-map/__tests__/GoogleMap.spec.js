import React from "react";
import { mount } from "enzyme";

import { GoogleMap } from "../GoogleMap";
import { createMapsMock } from "../../../mocks/MapsMock";
import GoogleMapEvents from "../GoogleMapEvents";

describe("GoogleMap", () => {
  let maps;

  beforeEach(() => {
    maps = createMapsMock();
  });

  describe("#componentDidMount", () => {
    it("should attach map to child div", () => {
      const wrapper = mount(
        <GoogleMap maps={maps} zoom={0} center={{ lat: 0, lng: 1 }} />,
      );

      const mapDiv = wrapper.find("div > div");

      expect(mapDiv.length).toBe(1);
      expect(maps.Map).toHaveBeenCalledTimes(1);
      expect(maps.Map).toHaveBeenLastCalledWith(mapDiv.getDOMNode());
    });

    it("should pass default options to map", () => {
      mount(<GoogleMap maps={maps} zoom={0} center={{ lat: 0, lng: 1 }} />);

      expect(maps.Map).toHaveBeenCalledTimes(1);

      const setValuesMock = maps.Map.mock.instances[0].setValues;

      expect(setValuesMock).toHaveBeenCalledTimes(1);
      expect(setValuesMock).toHaveBeenLastCalledWith({
        zoom: 0,
        center: { lat: 0, lng: 1 },
        mapTypeId: "ROADMAP",
        clickableIcons: true,
        disableDefaultUI: true,
        disableDoubleClickZoom: false,
      });
    });

    it("should pass options to map", () => {
      mount(
        <GoogleMap
          maps={maps}
          zoom={0}
          center={{ lat: 0, lng: 1 }}
          mapTypeId="HYBRID"
          clickableIcons={false}
          disableDoubleClickZoom
        />,
      );

      expect(maps.Map).toHaveBeenCalledTimes(1);

      const setValuesMock = maps.Map.mock.instances[0].setValues;

      expect(setValuesMock).toHaveBeenCalledTimes(1);
      expect(setValuesMock).toHaveBeenLastCalledWith({
        zoom: 0,
        center: { lat: 0, lng: 1 },
        mapTypeId: "HYBRID",
        clickableIcons: false,
        disableDefaultUI: true,
        disableDoubleClickZoom: true,
      });
    });

    it("should ignore unknown options", () => {
      mount(
        <GoogleMap
          maps={maps}
          zoom={0}
          center={{ lat: 0, lng: 1 }}
          foo="bar"
          baz={false}
        />,
      );

      expect(maps.Map).toHaveBeenCalledTimes(1);

      const setValuesMock = maps.Map.mock.instances[0].setValues;

      expect(setValuesMock).toHaveBeenCalledTimes(1);
      expect(setValuesMock).toHaveBeenLastCalledWith({
        zoom: 0,
        center: { lat: 0, lng: 1 },
        mapTypeId: "ROADMAP",
        clickableIcons: true,
        disableDefaultUI: true,
        disableDoubleClickZoom: false,
      });
    });

    it("should attach handlers", () => {
      const handlers = Object.keys(GoogleMapEvents).reduce((acc, x) => {
        acc[x] = jest.fn();

        return acc;
      }, {});

      mount(
        <GoogleMap
          zoom={0}
          maps={maps}
          {...handlers}
          center={{ lat: 0, lng: 1 }}
        />,
      );

      expect(maps.Map).toHaveBeenCalledTimes(1);

      const { listeners } = maps.Map.mock.instances[0];

      expect(listeners).toMatchSnapshot();

      Object.keys(listeners).forEach(x => {
        const listener = listeners[x][0];

        listener(x);
      });

      Object.keys(GoogleMapEvents).forEach(handler => {
        const event = GoogleMapEvents[handler];

        expect(handlers[handler]).toHaveBeenCalledTimes(1);
        expect(handlers[handler]).toHaveBeenLastCalledWith(event);
      });
    });
  });

  describe("#componentDidUpdate", () => {
    it("pass only changed options to map", () => {
      const options = {
        zoom: 0,
        center: { lat: 0, lng: 1 },
        disableDefaultUI: true,
        mapTypeId: "HYBRID",
        clickableIcons: false,
        disableDoubleClickZoom: false,
      };

      const wrapper = mount(<GoogleMap {...options} maps={maps} />);

      expect(maps.Map.mock.instances.length).toBe(1);

      const setValuesMock = maps.Map.mock.instances[0].setValues;

      expect(setValuesMock).toHaveBeenCalledTimes(1);
      expect(setValuesMock).toHaveBeenLastCalledWith(options);

      wrapper.setProps({ disableDoubleClickZoom: true });

      expect(setValuesMock).toHaveBeenCalledTimes(2);
      expect(setValuesMock).toHaveBeenLastCalledWith({
        disableDoubleClickZoom: true,
      });

      wrapper.setProps({ disableDoubleClickZoom: true });

      expect(setValuesMock).toHaveBeenCalledTimes(2);

      wrapper.setProps({ clickableIcons: true });

      expect(setValuesMock).toHaveBeenCalledTimes(3);
      expect(setValuesMock).toHaveBeenLastCalledWith({ clickableIcons: true });
    });
  });

  describe("componentWillUnmount", () => {
    it("should remove all listeners on unmount", () => {
      const wrapper = mount(
        <GoogleMap maps={maps} zoom={0} center={{ lat: 0, lng: 1 }} />,
      );

      expect(maps.Map).toHaveBeenCalledTimes(1);
      expect(maps.event.clearInstanceListeners).toHaveBeenCalledTimes(0);

      wrapper.unmount();

      expect(maps.event.clearInstanceListeners).toHaveBeenCalledTimes(1);
    });
  });
});
