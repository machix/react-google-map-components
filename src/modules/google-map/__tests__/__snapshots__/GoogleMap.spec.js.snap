// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleMap #componentDidMount should attach handlers 1`] = `
Object {
  "bounds_changed": Array [
    [Function],
  ],
  "center_changed": Array [
    [Function],
  ],
  "click": Array [
    [Function],
  ],
  "dblclick": Array [
    [Function],
  ],
  "drag": Array [
    [Function],
  ],
  "dragend": Array [
    [Function],
  ],
  "dragstart": Array [
    [Function],
  ],
  "heading_changed": Array [
    [Function],
  ],
  "idle": Array [
    [Function],
  ],
  "maptypeid_changed": Array [
    [Function],
  ],
  "mousemove": Array [
    [Function],
  ],
  "mouseout": Array [
    [Function],
  ],
  "mouseover": Array [
    [Function],
  ],
  "projection_changed": Array [
    [Function],
  ],
  "rightclick": Array [
    [Function],
  ],
  "tilesloaded": Array [
    [Function],
  ],
  "tilt_changed": Array [
    [Function],
  ],
  "zoom_changed": Array [
    [Function],
  ],
}
`;
